### PR TITLE
Bloom Filter implementation

### PR DESCRIPTION
**Pull request by mauris**
This should fulfill #15. 

The only thing I'd like some discussion for is:

- I have a set of 5 default hash functions and allowed user to introduce their own hash functions (as seen in the test). These hash functions are using the universal independent hash functions.
- `contains()` only returns a probability that an item is in the set - it does not 100% guarantee that the item is in the set. Rather, we can guarantee that an item is NOT in the set. So this leaves room for discussion if we should name it `probablyContains()`, or introduce another method `notContains()`, or leave it as it is to conform to how other methods are named.

### CODE REVIEW

File: lib/data-structures/BloomFilter.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
+/*
+  This is the default hash functions we would like to provide.
+  It may not be all-encompassing, but it's better than none.
+  All numbers used are prime numbers.
+*/
+// (ax + b) % p
+const defaultHashFunctions = [
+  (x) => (25 * x + 13) % 31,
+  (x) => (109 & x + 71) % 139,
```
Comment (yangshun) [inline lib/data-structures/BloomFilter.js]: Is this meant to be a *?

File: lib/data-structures/BloomFilter.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+class BloomFilter {
+  /**
+   * Initialize a bloom filter with an appropriate size
+   * and array of hash functions.
+   * @param {number} size The size of bloom filter to initialize as.
+   * @param {*[]} hashFunctions The array of hash functions to use.
+   */
+  constructor(size, hashFunctions) {
+    this._bits = Array(size).fill(false);
```
Comment (yangshun) [inline lib/data-structures/BloomFilter.js]: Should we specify a default size?

File: lib/data-structures/__tests__/BloomFilter.test.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    expect(bf._hashFunctions.length).toBe(5);
+  });
+
+  describe('add()', () => {
+    test('using numbers only', () => {
+      const bf = new BloomFilter(20);
+      bf.add(5);
+
+      // expect some bits to be set to true
+      expect(bf._bits.filter((bit) => bit === true).length).toBeGreaterThan(0);
```
Comment (yangshun) [inline lib/data-structures/__tests__/BloomFilter.test.js]: `(bit) => bit === true` can be shortened to `(bit) => bit`

File: lib/data-structures/__tests__/BloomFilter.test.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import BloomFilter from '../BloomFilter';
+
+const objectHashFunctions = [
+  (x) => (25 * x.value + 13) % 31,
+  (x) => (109 & x.value + 71) % 139,
```
Comment (yangshun) [inline lib/data-structures/__tests__/BloomFilter.test.js]: Same comment here

File: lib/data-structures/BloomFilter.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    } else {
+      this._hashFunctions = defaultHashFunctions;
+    }
+  }
+
+  /**
+   * Add an item to the bloom filter to register its membership.
+   * @param {*} item The item to be added.
+   */
+  add(item) {
```
Comment (yangshun) [inline lib/data-structures/BloomFilter.js:36]: Eventually we wanna support beyond integers. Could you help to add a TODO?

File: lib/data-structures/BloomFilter.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+   * @param {*} item The item whose membership is to be tested.
+   * @return {boolean} Whether the item is probably in the set or not.
+   *        If a false is returned, the item is surely not in the set.
+   */
+  contains(item) {
+    let result = true;
+
+    this._hashFunctions
+      .forEach((hashFunction) => {
+        result = result && this._bits[hashFunction(item) % this._size];
```
Comment (yangshun) [inline lib/data-structures/BloomFilter.js]: I think you can do this:

```js
return this._hashFunctions.every((hashFunction) => this._bits[hashFunction(item) % this._size]);
```

File: lib/data-structures/__tests__/BloomFilter.test.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+      // expect some bits to be set to true
+      expect(bf._bits.filter((bit) => bit === true).length).toBeGreaterThan(0);
+    });
+  });
+
+  describe('contains()', () => {
+    test('using numbers only', () => {
+      const bf = new BloomFilter(20);
+      bf.add(5);
```
Comment (yangshun) [inline lib/data-structures/__tests__/BloomFilter.test.js:41]: Can we add a few more tests for positive and negative cases?

File: lib/data-structures/BloomFilter.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  (x) => (547 * x + 383) % 997,
+  (x) => (173 * x + 149) % 499
+];
+
+class BloomFilter {
+  /**
+   * Initialize a bloom filter with an appropriate size
+   * and array of hash functions.
+   * @param {number} size The size of bloom filter to initialize as.
+   * @param {*[]} hashFunctions The array of hash functions to use.
```
Comment (yangshun) [inline lib/data-structures/BloomFilter.js]: I think it should be `{Function[]}`.

**Review by yangshun — COMMENTED** [review]
Just a couple of nits 😄

**Comment by yangshun** [comment]
Thanks, this was fast and awesome!

File: lib/data-structures/BloomFilter.js (hunk 1 above)
Comment (mauris) [reply lib/data-structures/BloomFilter.js]: wah yes. Not sure how it got there. will fix

File: lib/data-structures/BloomFilter.js (hunk 8 above)
Comment (mauris) [reply lib/data-structures/BloomFilter.js]: Okay I'll put `{Function[]}`.

File: lib/data-structures/BloomFilter.js (hunk 2 above)
Comment (mauris) [reply lib/data-structures/BloomFilter.js]: The size really depends on the number of items the user expects and the values of the items. Hard to give a ballpark figure here. I'd think it's already considered bad to specify a default set of independent hash functions

File: lib/data-structures/BloomFilter.js (hunk 5 above)
Comment (mauris) [reply lib/data-structures/BloomFilter.js:36]: Hmm I've not indicated that `item` is an integer. In the tests I've also introduced adding objects as `item` in the `add` method.

File: lib/data-structures/BloomFilter.js (hunk 6 above)
Comment (mauris) [reply lib/data-structures/BloomFilter.js]: will change

File: lib/data-structures/__tests__/BloomFilter.test.js (hunk 3 above)
Comment (mauris) [reply lib/data-structures/__tests__/BloomFilter.test.js]: testing for type equality, but sure.

**Comment by yangshun** [comment]
Cool, just merge it when you feel it's ok! (:
